##### LIBS ###
## Stringr
# String operations
library(stringr)

## Preprocess Core
# quantile normalization
library(preprocessCore)
library(reshape2)
library(ggfortify)
library(car)
library(DMwR)
##### DATA ###
## Intron counts file
# Intron counts per sample, generated by Leafcutter.
intron_counts <- read.table("~/Dropbox/Erik Schutte Internship 2016/Data/Leafcutter/gsTcell_perind_numers.counts",
                            header = T, sep=" ", dec=".")

##### PREP ###
## Column names
# Adjust column names, so they are more human readable.
new.colnames <- lapply( colnames(intron_counts), function(colname) {
  str_split(string = colname, pattern = "_[0-9]{6}")[[1]][1]
})
colnames(intron_counts) <- new.colnames

## Standardize values
# Standardize values across individuals for each intron.
scaled_across_indv <- t(scale(t(intron_counts)))

## Quantile normalization
# Normalize quantiles across introns.
norm_intron_counts <- normalize.quantiles(scaled_across_indv, copy = T)
colnames(norm_intron_counts) <- colnames(intron_counts) # Set column names from intron_counts
rownames(norm_intron_counts) <- rownames(intron_counts) # Set row names from intron_counts

plot(density(norm_intron_counts[2,]))
boxplot(norm_intron_counts[1:5000,])



pca <- prcomp(t(norm_intron_counts))

plot(pca$x)

# Filter out known faulty samples .
norm_intron_counts <- norm_intron_counts[,-grep("batch2_TCC.02.*",colnames(norm_intron_counts))] # -> e.g batch 2 tcc 02 only has 3 timepoints measured.
norm_intron_counts <- norm_intron_counts[,-grep("batch[0-9]+_TCC[0-9]+.*",colnames(norm_intron_counts))] # -> norway samples have a different format, they should be removed.
norm_intron_counts <- norm_intron_counts[,-grep("batch4_TCC.17.*", colnames(norm_intron_counts))] # -> batch 17 only has 3 itmepoints measured.

#Rename samples names to match samples from snp data so we can compare which are there and which arent. still 2 samples that have to be left out.
# names <- sapply(colnames(norm_intron_counts), gsub, pattern="\\.", replacement="_", USE.NAMES = F) 
# names <- sapply(names, gsub, pattern="batch[0-9]_", replacement="", USE.NAMES = F)
# names <- gsub(names, pattern="_t[0-9]+", replacement="")


# Write scaled quantile normalized data to tsv.
write.table(norm_intron_counts, "~/Dropbox/Erik Schutte Internship 2016/Data/Leafcutter/intron_counts_scaled_qnorm.tsv",sep="\t")

## Intron `BED-like` file
# Creates a file that will fulfill the role of gene.bed in MatrixEQTL analaysis.
intron_clusters <- rownames(norm_intron_counts)
intron.bed.location <- "~/Dropbox/Erik Schutte Internship 2016/Data/Leafcutter/intron.bed"
if ( file.exists( intron.bed.location ) ) {
  file.remove(intron.bed.location) # Because we append to the file, we want to remove it everytime we re-run this script.
}
lapply( str_split(intron_clusters, ":"), function(intron) {
  chr <- str_split(intron[1],"chr")[[1]][2] # Remove the chr before the chromosome indication.
  if (chr == "X") {
    chr <- 23 # If chromsome is X, create integer identifier as 23
  } else if (chr == "Y") {
    chr <- 24 # If chromsome is Y, create integer identifier as 23
  }
  start <- intron[2]
  end <- intron[3]
  clu <- str_split(intron[4],"_")[[1]][2] # Remove the Clu_ in front of the cluster identifier.
  id <- paste(clu,":",chr,":",start,":",end,sep="")
  
  write.table(x = paste(id, chr, start, end), file = intron.bed.location,
              append = T, sep = "\t", quote = F, col.names = F, row.names = F)
})


### Visual only, may adjust later depending on ordor of scaling log an transformation
par(mfrow=c(3,3))
plot(density(scaled_across_indv[1,]), main = "scaled")
hist(scaled_across_indv[1,], main = "scaled")
plot(density(scaled_across_indv.log[1,]), main = "log scaled")
hist(scaled_across_indv.log[1,], main = "log scaled")
plot(density(norm_intron_counts[1,]), main = "scaled normalized")
hist(norm_intron_counts[1,], main = "scaled normalized")
plot(density(norm_intron_counts.log[1,]), main = "log scaled normalized")
hist(norm_intron_counts.log[1,], main = "log scaled normalized")
plot(density(log2(norm_intron_counts[1,]+1)), main = "scaled normalized log")

plot(density(scaled_across_indv[10,]), main = "scaled")
hist(scaled_across_indv[10,], main = "scaled")
plot(density(scaled_across_indv.log[10,]), main = "log scaled")
hist(scaled_across_indv.log[10,], main = "log scaled")
plot(density(norm_intron_counts[10,]), main = "scaled normalized")
hist(norm_intron_counts[10,], main = "scaled normalized")
plot(density(norm_intron_counts.log[10,]), main = "log scaled normalized")
hist(norm_intron_counts.log[10,], main = "log scaled normalized")
plot(density(log2(norm_intron_counts[10,]+2)), main = "scaled normalized log")